packages/lesswrong/server/editor/conversionUtils.ts

```diff
@@ -0,0 +1,286 @@
+}
+
+export async function markdownToHtml(markdown: string): Promise<string> {
+  const html = markdownToHtmlNoLaTeX(markdown)
+  return await mjPagePromise(html, trimLatexAndAddCSS)
+}
+
+export async function ckEditorMarkupToHtml(markup: string): Promise<string> {
+  // Sanitized CKEditor markup is just html
+  const html = sanitize(markup)
+  const trimmedHtml = trimLeadingAndTrailingWhiteSpace(html)
+  // Render any LaTeX tags we might have in the HTML
+  return await mjPagePromise(trimmedHtml, trimLatexAndAddCSS)
+}
+
+export async function dataToHTML(data, type, sanitizeData = false) {
+  switch (type) {
+    case "html":
+      return sanitizeData ? sanitize(data) : await mjPagePromise(data, trimLatexAndAddCSS)
+    case "ckEditorMarkup":
+      return await ckEditorMarkupToHtml(data)
+    case "draftJS":
+      return await draftJSToHtmlWithLatex(data)
+    case "markdown":
+      return await markdownToHtml(data)
+    default: throw new Error(`Unrecognized format: ${type}`);
+  }
+}
+
+export function dataToMarkdown(data, type) {
+  if (!data) return ""
+  switch (type) {
+    case "markdown": {
+      return data
+    }
+    case "html": {
+      return htmlToMarkdown(data)
+    }
+    case "ckEditorMarkup": {
+      return ckEditorMarkupToMarkdown(data)
+    }
+    case "draftJS": {
+      try {
+        const contentState = convertFromRaw(data);
+        const html = draftToHTML(contentState)
+        return htmlToMarkdown(html)
+      } catch(e) {
+        // eslint-disable-next-line no-console
+        console.error(e)
+      }
+      return ""
+    }
+    default: throw new Error(`Unrecognized format: ${type}`);
+  }
+}
+
+export async function dataToCkEditor(data, type) {
+  switch (type) {
+    case "html":
+      return sanitize(data);
+    case "ckEditorMarkup":
+      return data;
+    case "draftJS":
+      return await draftJSToHtmlWithLatex(data);
+    case "markdown":
+      return await markdownToHtml(data)
+    default: throw new Error(`Unrecognized format: ${type}`);
+  }
+}
+
+/**
+ * When we calculate the word count we want to ignore footnotes. There's two syntaxes
+ * for footnotes in markdown:
+ *
+ * 1.  ^**[^](#fnreflexzxp4wr9h)**^
+ *
+ *     The contents of my footnote
+ *
+ * and
+ *
+ * [^1]: The contents of my footnote.
+ *
+ * In both cases, the footnote must start at character 0 on the line. The strategy here
+ * is just to find the first place where this occurs and then to ignore to the end of
+ * the document.
+ */
+export async function dataToWordCount(data, type) {
+  try {
+    const markdown = dataToMarkdown(data, type) ?? "";
+    const withoutFootnotes = markdown
+      .split(/^1\. {2}\^\*\*\[\^\]\(#(.|\n)*/m)[0]
+      .split(/^\[\^1\]:.*/m)[0];
+    const words = withoutFootnotes.match(/[^\s]+/g) ?? [];
+    return words.length;
+  } catch(err) {
+    // eslint-disable-next-line no-console
+    console.error("Error in dataToWordCount", data, type, err)
+    return 0
+  }
+}
```

Comment:
Is this all copied from `make_editable_callbacks` or other places too?

Reply:
Yep - I just moved it to avoid a dependency cycle.